src/mmw/apps/modeling/geoprocessing.py

```diff
@@ -60,29 +61,36 @@ def sjs_submit(host, port, args, data, retry=None):
     """
     base_url = 'http://{}:{}'.format(host, port)
     jobs_url = '{}/jobs?{}'.format(base_url, args)
-    response = requests.post(jobs_url, data=json.dumps(data))
+
+    try:
+        response = requests.post(jobs_url, data=json.dumps(data))
+    except ConnectionError as exc:
+        if retry is not None:
+            retry(exc=exc)
 
     if response.ok:
         job = response.json()
     else:
         error = response.json()
-        if error['status'] == 'NO SLOTS AVAILABLE' and retry:
-            try:
-                retry()
-            except MaxRetriesExceededError:
-                raise Exception('No slots available in Spark JobServer.\n'
-                                'Details = {}'.format(response.text))
+
+        if error['status'] == 'NO SLOTS AVAILABLE' and retry is not None:
+            retry(exc=Exception('No slots available in Spark JobServer.\n'
+                                'Details = {}'.format(response.text)))
         elif error['result'] == 'context geoprocessing not found':
             reboot_sjs_url = '{}/contexts?reset=reboot'.format(base_url)
             context_response = requests.put(reboot_sjs_url)
 
             if context_response.ok:
-                if retry:
-                    retry()
+                if retry is not None:
+                    retry(exc=Exception('Geoprocessing context missing in '
```

Comment:
Yeah. A caveat I encountered is:

> If this argument is set and retry is called while an exception was raised (sys.exc_info() is set) it will attempt to reraise the current exception.
> If no exception was raised it will raise the exc argument provided.